intermediate_source/custom_function_conv_bn_tutorial.py

```diff
@@ -0,0 +1,346 @@
+# and BatchNorm2D. Eventually we'd chain them together in our unified
+# backward function, but below we first implement them as their own
+# custom functions so we can validate their correctness individually
+import torch
+from torch.autograd.function import once_differentiable
+import torch.nn.functional as F
+
+def convolution_backward(grad_out, X, weight):
+    grad_input = F.conv2d(X.transpose(0, 1), grad_out.transpose(0, 1)).transpose(0, 1)
+    grad_X = F.conv_transpose2d(grad_out, weight)
+    return grad_X, grad_input
+
+class Conv2D(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, X, weight):
+        ctx.save_for_backward(X, weight)
+        return F.conv2d(X, weight)
+
+    # Use @once_differentiable by default unless we intend to double
+    # backward
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, grad_out):
+        X, weight = ctx.saved_tensors
+        return convolution_backward(grad_out, X, weight)
+
+######################################################################
+# When testing with gradcheck, it is important to use double precision
+weight = torch.rand(5, 3, 3, 3, requires_grad=True, dtype=torch.double)
+X = torch.rand(10, 3, 7, 7, requires_grad=True, dtype=torch.double)
+torch.autograd.gradcheck(Conv2D.apply, (X, weight))
+
+######################################################################
+# Batch Norm has two modes: training and eval mode. In training mode
+# the sample statistics are a function of the inputs, in eval mode,
+# we use the saved running statistics, which are not a function of the inputs
+# This makes non-training mode's backward significantly simpler. Below
+# We implement and test only the training mode case.
+def unsqueeze_all(t):
+    # Helper function to unsqueeze all the dimensions that we reduce over
+    return t[None, :, None, None]
+
+def batch_norm_backward(grad_out, X, sum, sqrt_var, N, eps):
+    # We use the formula: out = (X - mean(X)) / (sqrt(var(X)) + eps)
+    # in batch norm 2d's forward. To simplify our derivation, we follow the
+    # chain rule and compute the gradients as follows before accumulating
+    # them all into a final grad_input.
+    #  1) 'grad of out wrt var(X)' * 'grad of var(X) wrt X'
+    #  2) 'grad of out wrt mean(X)' * 'grad of mean(X) wrt X'
+    #  3) 'grad of out wrt X in the numerator' * 'grad of X wrt X'
+    # We then rewrite the formulas to use as few extra buffers as possible
+    tmp = -((X - unsqueeze_all(sum) / N) * grad_out).sum(dim=(0, 2, 3))
+    d_denom = tmp / (sqrt_var + eps)**2  # d_denom = -num / denom**2
+    # It is useful to delete tensors when you no longer need them with `del`
+    # For example, we could've done `del tmp` here because we won't use it later
+    # In this case, it's not a big difference because tmp only has size of (C,)
+    # The important thing is avoid allocating NCHW-sized tensors unnecessarily
+    d_var = d_denom / (2 * sqrt_var)  # denom = torch.sqrt(var) + eps
```

Comment:
I am not sure these lines don't actually allocate extra buffers?
Note that this is not a very strict requirement, but we're trying to show what to do to save memory, so I think we should make sure we see the biggest benefit possible.

Reply:
I think this one does, but `d_denom` only has size `(C,)` vs something like grad_input, X, grad_out which have  `(N, C, H, W)` so I figured it would be a good way to demonstrate that distinction as well (see code comment above). I end up only allocating one extra NCHW-sized tensor in the end (in addition to grad_out and X), which should be the theoretical best we can do given that we should not modify X or grad_out in-place. Had to pull off this weird hack as well - see below in step (3).